lib/Models/CatalogGroup.js

```diff
@@ -412,4 +402,135 @@ CatalogGroup.prototype.sortItems = function(sortRecursively) {
+ *                                           to the given value.  If not specified, the property is left unchanged.
+ * @param {Object} [options.shareKeyIndex] An optional index of all catalog items by their share ids - this will be
+ *                                         used to match catalog items in itemsJson to their already-existing equivalents
+ *                                         even if those equivalents have been moved outside the catalogGroup specified.
+ * @param {CatalogGroup} catalogGroup The catalogGroup to update.
+ *
+ * @returns {Promise} A promise that resolves when the update is complete.
+ */
+CatalogGroup.updateItems = function (itemsJson, options, catalogGroup) {
+    if (!(itemsJson instanceof Array)) {
+        throw new DeveloperError('JSON catalog description must be an array of groups.');
+    }
+
+    options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+    var onlyUpdateExistingItems = defaultValue(options.onlyUpdateExistingItems, false);
+
+    var promises = [];
+
+    for (var itemIndex = 0; itemIndex < itemsJson.length; ++itemIndex) {
+        var item = itemsJson[itemIndex];
+
+        if (!defined(item.name) && !defined(item.id)) {
+            throw new RuntimeError('A catalog member must have a name or a id for matching.');
+        }
+
+        var existingItem;
+        if (item.id) {
+            if (options.shareKeyIndex) {
+                existingItem = options.shareKeyIndex[item.id];
+
+                if (existingItem && existingItem.parent !== catalogGroup && item.isEnabled && !existingItem.isEnabled) {
+                    // We are enabling an existing item that has been moved since our itemsJson was generated
+                    // (probably an old share link) - make sure all its parents are open so it's visible.
+                    existingItem.parent.openWithAllParents();
+                }
+            } else {
+                existingItem = catalogGroup.getChildForShareKey(item.id);
+            }
+        } else if (item.name) {
+            existingItem = catalogGroup.findFirstItemByName(item.name);
+        }
+
+        if (!defined(existingItem)) {
+            // Skip this item entirely if we're not allowed to create it.
+            if (onlyUpdateExistingItems) {
+                continue;
+            }
+
+            if (!defined(item.name)) {
+                throw new RuntimeError('A newly created catalog member must have a name.');
+            }
+
+            if (!defined(item.type)) {
+                throw new RuntimeError('A catalog member must have a type.');
+            }
+
+            existingItem = createCatalogMemberFromType(item.type, catalogGroup.terria);
+
+            catalogGroup.add(existingItem);
+        } else {
+            // We only want to use the id for matching, not overwrite it.
+            delete item.id;
```

Comment:
Can we avoid this delete somehow?  Maybe by explicitly skipping `id` in `updateFromJson`?  Two reasons:
* Deleting properties pretty destroys the browser's ability to optimize property access to this instance.  Maybe not a big deal here, but still...
* It's better to avoid mutating the parameters to a function whenever possible.